src/audio/eq_fir.c

```diff
@@ -63,20 +63,100 @@ struct comp_data {
 	uint32_t period_bytes;
 	int32_t *fir_delay;
 	size_t fir_delay_size;
+	void (*eq_fir_func_even)(struct fir_state_32x16 fir[],
+				 struct comp_buffer *source,
+				 struct comp_buffer *sink,
+				 int frames, int nch);
 	void (*eq_fir_func)(struct fir_state_32x16 fir[],
 			    struct comp_buffer *source,
 			    struct comp_buffer *sink,
 			    int frames, int nch);
-	void (*eq_fir_func_odd)(struct fir_state_32x16 fir[],
-				struct comp_buffer *source,
-				struct comp_buffer *sink,
-				int frames, int nch);
 };
 
-static void eq_fir_passthrough(struct fir_state_32x16 fir[],
-			       struct comp_buffer *source,
-			       struct comp_buffer *sink,
-			       int frames, int nch)
+/* The optimized FIR functions variants need to be updated into this
+ * function. The eq_fir_func is a function that can process any number
+ * of samples. The eq_fir_func_even is for optimized version that is
+ * guaranteed to be called with even samples number.
+ *
+ */
+static inline int eq_fir_select_fir_func(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct sof_ipc_comp_config *config = COMP_GET_CONFIG(dev);
+
+	switch (config->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		trace_eq("f16");
+#if FIR_GENERIC
+		cd->eq_fir_func_even = eq_fir_s16;
+		cd->eq_fir_func = eq_fir_s16;
+#endif
```

Comment:
sorry, I mean this function will call a 
```
static inline void set_fir_func(cd);
```
This function will be defined in condition code
```
#if HIFI3
static inline void set_fir_func(..)
{
}
#elif HIFI2
static inline void set_fir_func()
{
}
#else
/* generic */
static inline void set_fir_func()
{
}
\endif
```